src/test/java/com/aws/greengrass/componentmanager/builtins/S3DownloaderTest.java

```diff
@@ -173,4 +189,58 @@ void GIVEN_s3_artifact_uri_WHEN_error_in_getting_from_s3_THEN_fail() throws Exce
             ComponentTestResourceHelper.cleanDirectory(testCache);
         }
     }
+
+    @Test
+    void GIVEN_s3_artifact_uri_WHEN_cloud_deployment_error_in_getting_from_s3_THEN_retry(ExtensionContext context) throws Exception {
+        ignoreExceptionOfType(context, RetryableCloudDeploymentException.class);
+        Path testCache = ComponentTestResourceHelper.getPathForLocalTestCache();
+        Path artifactFilePath =
+                Files.write(tempDir.resolve("artifact.txt"), Collections.singletonList(VALID_ARTIFACT_CONTENT),
+                        StandardCharsets.UTF_8, StandardOpenOption.CREATE, StandardOpenOption.APPEND);
+        ResponseInputStream<GetObjectResponse> getObjectResponse = null;
+        try {
+            String checksum = Base64.getEncoder()
+                    .encodeToString(MessageDigest.getInstance("SHA-256").digest(Files.readAllBytes(artifactFilePath)));
+            Path saveToPath = testCache.resolve(TEST_COMPONENT_NAME).resolve(TEST_COMPONENT_VERSION);
+            if (Files.notExists(saveToPath)) {
+                Files.createDirectories(saveToPath);
+            }
+            S3Downloader s3Downloader = new S3Downloader(s3SdkClientFactory,
+                    new ComponentIdentifier(TEST_COMPONENT_NAME, new Semver(TEST_COMPONENT_VERSION)),
+                    ComponentArtifact.builder().artifactUri(new URI(VALID_ARTIFACT_URI))
+                            .checksum(checksum).algorithm(VALID_ALGORITHM).build(),
+                    saveToPath);
+            getObjectResponse = new ResponseInputStream<>(GetObjectResponse.builder().build(),
+                    AbortableInputStream.create(new ByteArrayInputStream(Files.readAllBytes(artifactFilePath))));
+            when(s3Client.getObject(any(GetObjectRequest.class))).thenThrow(S3Exception.builder().statusCode(500).build())
+                    .thenReturn(getObjectResponse);
+            HeadObjectResponse  headObjectResponse = HeadObjectResponse.builder()
+                    .contentLength(Files.size(artifactFilePath)).build();
+            when(s3Client.headObject(any(HeadObjectRequest.class))).thenReturn(headObjectResponse);
+
+            s3Downloader.download();
+            byte[] downloadedFile = Files.readAllBytes(saveToPath.resolve("artifact.txt"));
+            assertThat("Content of downloaded file should be same as the artifact content",
+                    Arrays.equals(Files.readAllBytes(artifactFilePath), downloadedFile));
```

Comment:
How are you verifying that retry happened before eventually getting a success response?

Reply:
I think you missed this one

Reply:
The response is returned at line 210, which is the response of GetObject API

Reply:
The goal of the test is to not verify the response though, it's to ensure that the correct number of retry attempts were made, and the way to verify that would be to check that the relevant API/method was called the appropriate number of times. Like you do in the very next test at line 249

Reply:
Got it, updated!